src/framework/results/data/data_container.hpp

```diff
@@ -0,0 +1,299 @@
+                            const std::string &memory,
+                            T &&datum, bool variance);
+  void add_average_snapshot(const std::string &type,
+                            const std::string &label,
+                            const std::string &memory,
+                            T &datum, bool variance);
+  void add_average_snapshot(const std::string &type,
+                            const std::string &label,
+                            const std::string &memory,
+                            const T &datum, bool variance);
+
+  //----------------------------------------------------------------
+  // Config
+  //----------------------------------------------------------------
+
+  // Enable data container
+  void enable(bool value) {enabled_ = value;}
+
+  // Empty engine of stored data
+  void clear();
+
+  // Combine engines for accumulating data
+  // Second engine should no longer be used after combining
+  // as this function should use move semantics to minimize copying
+  DataContainer<T> &combine(DataContainer<T> &&eng); // Move semantics
+  DataContainer<T> &combine(const DataContainer<T> &eng); // Copy semantics
+
+  // Operator overload for combine
+  // Note this operator is not defined to be const on the input argument
+  inline DataContainer<T> &operator+=(const DataContainer<T> &data) {
+    return combine(data);
+  }
+  inline DataContainer<T> &operator+=(DataContainer<T> &&data) {
+    return combine(std::move(data));
+  }
+
+  //----------------------------------------------------------------
+  // Data containers
+  //----------------------------------------------------------------
+
+  // Additional data
+  stringmap_t<T> additional_data_;
+
+  // Pershot snapshots
+  stringmap_t<PershotSnapshot<T>> pershot_snapshots_;
+
+  // Average snapshots
+  stringmap_t<AverageSnapshot<T>> average_snapshots_;
+
+  // Store whether this data type is enabled or not
+  bool enabled_ = true;
+};
+
+//============================================================================
+// Implementations
+//============================================================================
+
+//------------------------------------------------------------------
+// Add additional data
+//------------------------------------------------------------------
+
+template <typename T>
+void DataContainer<T>::add_additional_data(const std::string &key,
+                                           T &&data) {
+  if (enabled_) {
+    additional_data_[key] = std::move(data);
+  }
+}
+
+template <typename T>
+void DataContainer<T>::add_additional_data(const std::string &key,
+                                           const T &data) {
+  if (enabled_) {
+    additional_data_[key] = data;
+  }
+}
+
+template <typename T>
+void DataContainer<T>::add_additional_data(const std::string &key,
+                                           T &data) {
+  if (enabled_) {
+    additional_data_[key] = data;
+  }
+}
```

Comment:
Shouldn't this work having only the one accepting the `const T&` (same for `add_pershot_snapshot` and `add_average_snapshot`?  Am I missing something?

Reply:
I did some testing and I found i needed to add both `const T&` and `T&`, if one was missing that would get converted via the default json conversion `T&&` method in the derived class